File: lab3/src/cpp/src/hashtable.cpp
#include "hashtable.hpp"

#include <functional>
#include <sstream>

HashTable::HashTable(size_t initialCapacity) : m_capacity(initialCapacity) {
    m_buckets.resize(m_capacity, nullptr);
}

void HashTable::clear() noexcept {
    for (auto node : m_buckets) {
        while (node) {
            HashNode* next = node->next;
            delete node;
            node = next;
        }
    }
    m_buckets.clear();
    m_count = 0;
}

HashTable::~HashTable() {
    clear();
}

HashTable::HashTable(HashTable&& other) noexcept
    : m_buckets(std::move(other.m_buckets)), m_capacity(other.m_capacity), m_count(other.m_count) {
    other.m_capacity = 0;
    other.m_count = 0;
}

HashTable& HashTable::operator=(HashTable&& other) noexcept {
    if (this != &other) {
        clear();
        m_buckets = std::move(other.m_buckets);
        m_capacity = other.m_capacity;
        m_count = other.m_count;
        other.m_capacity = 0;
        other.m_count = 0;
    }
    return *this;
}

size_t HashTable::hashFunction(std::string_view key) const {
    return std::hash<std::string_view>{}(key) % m_capacity;
}

void HashTable::resize() {
    size_t oldCapacity = m_capacity;
    auto oldBuckets = std::move(m_buckets);

    m_capacity = oldCapacity * 2;
    m_buckets.assign(m_capacity, nullptr);
    m_count = 0;

    for (auto node : oldBuckets) {
        while (node) {
            insert(node->key, node->value);
            HashNode* next = node->next;
            delete node;
            node = next;
        }
    }
}

void HashTable::insert(std::string_view key, std::string_view value) {
    size_t index = hashFunction(key);
    HashNode* node = m_buckets[index];

    while (node) {
        if (node->key == key) {
            node->value = value;
            return;
        }
        node = node->next;
    }

    auto* newNode = new HashNode(key, value);
    newNode->next = m_buckets[index];
    m_buckets[index] = newNode;
    m_count++;

    if (static_cast<double>(m_count) / m_capacity > LOAD_FACTOR_THRESHOLD) {
        resize();
    }
}

bool HashTable::remove(std::string_view key) {
    size_t index = hashFunction(key);
    HashNode* node = m_buckets[index];
    HashNode* prev = nullptr;

    while (node) {
        if (node->key == key) {
            if (prev)
                prev->next = node->next;
            else
                m_buckets[index] = node->next;

            delete node;
            m_count--;
            return true;
        }
        prev = node;
        node = node->next;
    }
    return false;
}

std::string HashTable::get(std::string_view key) const {
    size_t index = hashFunction(key);
    HashNode* node = m_buckets[index];

    while (node) {
        if (node->key == key) return node->value;
        node = node->next;
    }
    return "";
}

bool HashTable::contains(std::string_view key) const {
    size_t index = hashFunction(key);
    HashNode* node = m_buckets[index];

    while (node) {
        if (node->key == key) return true;
        node = node->next;
    }
    return false;
}

std::string HashTable::print() const {
    std::ostringstream oss;
    oss << "{";
    bool first = true;
    for (const auto* node : m_buckets) {
        while (node) {
            if (!first) oss << ", ";
            oss << node->key << ": " << node->value;
            first = false;
            node = node->next;
        }
    }
    oss << "}";
    return oss.str();
}

File: lab3/src/go/internal/hashtable.go
package internal

import (
	"fmt"
	"strings"
)

type hashNode struct {
	key   string
	value string
	next  *hashNode
}

type HashEntry struct {
	Key   string
	Value string
}

type HashTable struct {
	buckets  []*hashNode
	capacity int
	count    int
}

const loadFactorThreshold = 0.75

func NewHashTable(initialCapacity int) *HashTable {
	if initialCapacity <= 0 {
		initialCapacity = 16
	}
	return &HashTable{
		buckets:  make([]*hashNode, initialCapacity),
		capacity: initialCapacity,
		count:    0,
	}
}

func (ht *HashTable) hashFunction(key string) int {
	hash := 0
	for i := 0; i < len(key); i++ {
		hash = 31*hash + int(key[i])
	}
	return (hash & 0x7FFFFFFF) % ht.capacity
}

func (ht *HashTable) Insert(key, value string) {
	index := ht.hashFunction(key)
	curr := ht.buckets[index]

	for curr != nil {
		if curr.key == key {
			curr.value = value
			return
		}
		curr = curr.next
	}

	newNode := &hashNode{key: key, value: value, next: ht.buckets[index]}
	ht.buckets[index] = newNode
	ht.count++

	if float64(ht.count)/float64(ht.capacity) > loadFactorThreshold {
		ht.resize()
	}
}

func (ht *HashTable) resize() {
	oldBuckets := ht.buckets
	ht.capacity *= 2
	ht.buckets = make([]*hashNode, ht.capacity)
	ht.count = 0

	for _, node := range oldBuckets {
		for node != nil {
			ht.Insert(node.key, node.value)
			node = node.next
		}
	}
}

func (ht *HashTable) Get(key string) (string, bool) {
	index := ht.hashFunction(key)
	curr := ht.buckets[index]
	for curr != nil {
		if curr.key == key {
			return curr.value, true
		}
		curr = curr.next
	}
	return "", false
}

func (ht *HashTable) Remove(key string) bool {
	index := ht.hashFunction(key)
	curr := ht.buckets[index]
	var prev *hashNode

	for curr != nil {
		if curr.key == key {
			if prev != nil {
				prev.next = curr.next
			} else {
				ht.buckets[index] = curr.next
			}
			ht.count--
			return true
		}
		prev = curr
		curr = curr.next
	}
	return false
}

func (ht *HashTable) GetAll() []HashEntry {
	entries := make([]HashEntry, 0, ht.count)
	for _, node := range ht.buckets {
		for node != nil {
			entries = append(entries, HashEntry{Key: node.key, Value: node.value})
			node = node.next
		}
	}
	return entries
}

func (ht *HashTable) Print() string {
	var sb strings.Builder
	sb.WriteString("{")
	first := true
	for _, node := range ht.buckets {
		for node != nil {
			if !first {
				sb.WriteString(", ")
			}
			fmt.Fprintf(&sb, "%s: %s", node.key, node.value)
			first = false
			node = node.next
		}
	}
	sb.WriteString("}")
	return sb.String()
}

func (ht *HashTable) GetCount() int {
	return ht.count
}